e2e-polybft/e2e/skyline_bridge_test.go

```diff
@@ -2061,3 +2064,225 @@ func TestE2E_SkylineBridge_Fund_Defund(t *testing.T) {
+	minterWalletCardano := apex.CardanoInfo.GenesisWallet
+
+	t.Run("Both directions 1000x 60min 90%", func(t *testing.T) {
+		const (
+			instances     = 1000
+			maxWaitTime   = 3600
+			successChance = 90 // 90%
+
+			// wait for tx timeout
+			numRetries = 500
+			waitTime   = time.Second * 10
+		)
+
+		sendAmount := new(big.Int).SetInt64(1_000_000)
+
+		type bridgingRequest struct {
+			src            cardanofw.ChainID
+			dest           cardanofw.ChainID
+			firstSenderIdx int
+			bridgingType   sendtx.BridgingType
+			receiver       *cardanofw.TestApexUser
+		}
+
+		bridgingRequests := []bridgingRequest{
+			{src: cardanofw.ChainIDPrime, dest: cardanofw.ChainIDCardano, firstSenderIdx: 0, bridgingType: sendtx.BridgingTypeNativeTokenOnSource, receiver: currencyReceiver},
+			{src: cardanofw.ChainIDCardano, dest: cardanofw.ChainIDPrime, firstSenderIdx: 0, bridgingType: sendtx.BridgingTypeNativeTokenOnSource, receiver: currencyReceiver},
+			{src: cardanofw.ChainIDPrime, dest: cardanofw.ChainIDCardano, firstSenderIdx: instances, bridgingType: sendtx.BridgingTypeCurrencyOnSource, receiver: nativeTokenReceiver},
+			{src: cardanofw.ChainIDCardano, dest: cardanofw.ChainIDPrime, firstSenderIdx: instances, bridgingType: sendtx.BridgingTypeCurrencyOnSource, receiver: nativeTokenReceiver},
+		}
+
+		seed := rand.Int63n(1_000_000_000)
+		r := rand.New(rand.NewSource(seed)) // New seeded random number generator
+
+		fmt.Printf("Test seed: %v\n", seed)
+
+		fmt.Printf("Funding users with native tokens\n")
+
+		for i := range instances {
+			_, err := cardanofw.FundUserWithToken(
+				ctx, apex, cardanofw.ChainIDPrime,
+				minterWalletPrime, apex.Users[i],
+				cardanofw.DefaultTokenName, cardanofw.DefaultTokenMintAmount,
+				uint64(1_100_000_000), uint64(2_500_000))
+			require.NoError(t, err)
+
+			_, err = cardanofw.FundUserWithToken(
+				ctx, apex, cardanofw.ChainIDCardano,
+				minterWalletCardano, apex.Users[i],
+				cardanofw.DefaultTokenName, cardanofw.DefaultTokenMintAmount,
+				uint64(1_100_000_000), uint64(2_500_000))
+			require.NoError(t, err)
+		}
+
+		fmt.Printf("Sending %v transactions in %v seconds\n", instances*len(bridgingRequests), maxWaitTime)
+
+		prevAmounts := make(map[int]map[string]*big.Int)
+		expectedAmounts := make(map[int]*big.Int)
+
+		var wg sync.WaitGroup
+
+		for brIdx, br := range bridgingRequests {
+			var err error
+
+			succeededCount := int64(0)
+
+			prevAmounts[brIdx], err = apex.GetBalance(ctx, br.receiver, br.dest)
+			require.NoError(t, err)
+
+			var tokenName string
+
+			if br.bridgingType == sendtx.BridgingTypeNativeTokenOnSource {
+				tokenName = wallet.AdaTokenName
+			} else {
+				tokenName = apex.GetTokenNameForChains(br.dest, br.src)
+			}
+
+			if amount, ok := prevAmounts[brIdx][tokenName]; ok {
+				expectedAmounts[brIdx] = new(big.Int).Set(amount)
+			} else {
+				expectedAmounts[brIdx] = big.NewInt(0)
+			}
+
+			for i := 0; i < instances; i++ {
+				success := successChance > r.Intn(100)
```

Comment:
I am not sure if we should keep test nondeterministic, but lets keep this random calculation for now